### PR TITLE
[CALCITE-1301] Add cancel flag to AvaticaStatement

### PR DESCRIPTION
**Pull request by julianhyde**
Allow connection properties that are enums to declare their value
type. (Extends the work done in [CALCITE-1263].)

### CODE REVIEW

File: avatica/core/src/main/java/org/apache/calcite/avatica/AvaticaConnection.java [hunk 1]
```diff
@@ -664,6 +665,13 @@ private int unbox(Integer i, int defaultValue) {
     return meta.connectionSync(handle, new ConnectionPropertiesImpl());
   }
 
+  /** Returns or creates a slot whose state can be changed to cancel a
+   * statement. */
+  public AtomicBoolean getCancelFlag(Meta.StatementHandle h)
+      throws NoSuchStatementException {
+    return new AtomicBoolean();
```
Comment (joshelser) [inline avatica/core/src/main/java/org/apache/calcite/avatica/AvaticaConnection.java]: Isn't this a worry if, by some means, we call `connection.getCancelFlag(h)` for the same `h` multiple times and end up with multiple instances? Can we initialize an `AtomicBoolean` on the `StatementHandle` itself?

File: avatica/core/src/main/java/org/apache/calcite/avatica/ConnectionProperty.java [hunk 2]
```diff
@@ -57,6 +57,24 @@
     ENUM,
     PLUGIN;
 
+    /** Deduces the class of a property of this type, given the default value
+     * and the user-specified value class (each of which may be null, unless
+     * this is an enum or a plugin). */
+    public Class deduceValueClass(Object defaultValue, Class valueClass) {
```
Comment (joshelser) [inline avatica/core/src/main/java/org/apache/calcite/avatica/ConnectionProperty.java]: These changes seem unrelated. Did they accidentally make it in?

File: avatica/core/src/test/java/org/apache/calcite/avatica/test/AvaticaUtilsTest.java [hunk 3]
```diff
@@ -174,6 +174,22 @@
           is("wrong value class; expected " + Size.class));
     }
 
+    // An enum with a default value that is an anonymous enum,
+    // not specifying value type.
+    final ConnectionPropertyImpl v = new ConnectionPropertyImpl("V",
```
Comment (joshelser) [inline avatica/core/src/test/java/org/apache/calcite/avatica/test/AvaticaUtilsTest.java]: Same comment here about unrelated changes.

**Comment by julianhyde** [comment]
Has been partially fixed in 92e6494, and what's left is covered by https://issues.apache.org/jira/browse/CALCITE-1361.
